File: Python/1059-prime_factors.py
from collections import Counter


PRIMES = set()


def factorize(num):
    if num in PRIMES:
        RES[num] += 1
        return
    i = 2
    while i * i <= num:
        if num % i == 0:
            factorize(i)
            factorize(num // i)
            return
        i += 1
    PRIMES.add(num)
    RES[num] += 1


if __name__ == '__main__':
    num = int(input())
    RES = Counter()
    factorize(num)
    ans = []
    for key in sorted(RES.keys()):
        expo = RES[key]
        if expo > 1:
            ans.append(f'{key}^{expo}')
        else:
            ans.append(str(key))
    print(num, '=', '*'.join(ans), sep='')

File: C++/1059-prime_factors.cc
#include <iostream>
#include <map>
#include <set>


std::set<long int> primes;


void factorize(long int num, std::map<long int, int> &res) {
    auto iter = primes.find(num);
    if (iter != primes.end()) {
        res[num] += 1;
        return;
    }
    for (int i = 2; i * i <= num; ++i) {
        if (num % i == 0) {
            factorize(i, res);
            factorize(num / i, res);
            return;
        }
    }
    primes.insert(num);
    res[num] += 1;
}


int main() {

    long int num;
    std::cin >> num;

    std::map<long int, int> res;
    factorize(num, res);

    std::cout << num << '=';
    auto iter = res.begin();
    std::cout << iter->first;
    if (iter->second > 1) {
        std::cout << '^' << iter->second;
    }
    for (++iter; iter!=res.end(); ++iter) {
        std::cout << '*' << iter->first;
        if (iter->second > 1) {
            std::cout << '^' << iter->second;
        }
    }
    std::cout << '\n';

    return 0;
}
